Refactor tf, add plan workflow

daniellohrey-hai opened the pull request:
## What

Update structure of Terraform directories, add GitHub Actions workflow to run `terraform plan`, update Makefile to use new structure.

Details:
- Added repo name to cookiecutter variables (in case it's different)
- Added github workflow to comment tf plan onto PRs (which references this guy: https://github.com/harrison-ai/actions-library/blob/main/.github/workflows/terraform-plan-env.yml - a starter actions-library repo)
- Removed docker compose / .env to run terraform
- Refactored terraform code to all live in /tf and use dynamically passed backend/var files based on env
- Updated Makefile to reflect these changes

## Why

Incorporating some ideas found in https://harrison-ai.atlassian.net/wiki/spaces/HSWE/pages/1934491649/Terraform+Workflow+Structure+Discussion

Hightlights:
- Adding automation
- Simplification
- Laying foundation for other improvements
- Some people have issues with docker on M1 chips

## To Do

- Fill out Makefile
- Update tests
- Update README / add docs
- Get roasted
- Incorporate feedback
- Agree on workflow pattern

schattingh reviewed (commented):
Hi @daniellohrey-hai 

As a starting point, can you please motivate why you are requesting each one of the changes in this PR?  What is the reason for each change?

Secondly, my gut reaction to the public `actions-library` repo is that its too easy for something that shouldn't be public to make its way in there unintended.  Not saying we need to remove it - will get back to you on it.

comozo commented on {{ cookiecutter.project_slug }}/tf/provider.tf:
```diff
@@ -11,14 +12,12 @@ terraform {
 provider "aws" {
-  region  = local.region
-  profile = local.profile
+  region  =  "ap-southeast-2"
```
What is the rationale of not using a variable here ?

comozo commented on {{ cookiecutter.project_slug }}/tf/provider.tf:
```diff
@@ -11,14 +12,12 @@ terraform {
   default_tags {
     tags = {
-      # NB: Capitalization is correct and required
-      Project = local.project
```
Did you find the project tag not useful ?

daniellohrey-hai replied:
It was essentially a duplicate of repo, although I did add another option in the config to differentiate, don't really have an opinion on this

comozo commented on {{ cookiecutter.project_slug }}/Makefile:
```diff
@@ -1,75 +1,7 @@
-include .env
-endif
+init:
+	terraform -chdir=tf init -backend-config=$(ENV).tfbackend -reconfigure
```
I'm not sure we can ditch docker-compose all together yet. They are folks running Mac/Arm64, windows,  Linux. The current architecture implements pattern called the 3 musketeers. I'll advise to have a look https://3musketeers.io/ it will give you more context I think.

daniellohrey-hai replied:
Am familiar with 3 musketeers
Does anyone have trouble running terraform itself? I think running docker compose here just adds extra steps/complexity without much benefit (and isn't even used in the pipeline which defeats the purpose it's used for in 3 musketeers)
Some people have also reported having issues with docker running on the M1 chips on the newer macs
If anyone can't run terraform itself it could be worth using docker, but I don't think we need compose to do that either way

schattingh replied:
@daniellohrey-hai you've mentioned a few time how docker-compose adds complexity?  Can you please explain how?  I have to admit I don't see how it does.

A hypothetical example:
```
docker compose run --rm terraform apply
```
vs
```
docker run -v ${PWD}:/app -e AWS_REGION=ap-southeast-2 -e ENVIRONMENT=prod terraform apply
```

Also bear in mind Terraform will always be run locally and in CI, so the repo needs to cater for both, as it currently does.

I've heard of folks having issues with M1 and Docker with some of the more exotic libraries we use, but not with any IaC tooling like Terraform.  Can you let me know who in the Org is having issues with this?

I'm not equating docker compose to 3 musketeers.  Happy to ditch the latter if we feel it's no longer needed.

daniellohrey-hai replied:
Haven't looked too deeply into the M1 issue, but fair enough.

Currently running plan locally looks like:
```
plan: .env
	docker-compose run --rm envvars ensure --tags terraform
	docker-compose run --rm --workdir /app/$(DIR) terraform terraform plan
```
which is fine, the Makefile abstracts the length of of any commands.

My issue is with `envvars.yml` and `dockercompose.yml` being extra files, and steps that can go wrong (that are difficult to debug). Seems unnecessary to me, but if you think it's needed then happy to leave it.

Is the plan to run docker in CI as well?
And why is running docker locally necessary? have people reported issues running terraform?

daniellohrey-hai replied:
I'm not actually super concerned where we end up with a lot of this, as long as we make a thought-out decision and stick with it; my main concern is having things in a state where we can start to work on some reusable automation (across Harrison at least).

schattingh replied:
> And why is running docker locally necessary? have people reported issues running terraform?

There are many reasons to run Terraform locally as well as in CI.  We pin to specific versions of Terraform.  With a number of repos in existence, this isn't the same across all repos.  Juggling these versions outside of docker?  No thanks.

Three people (myself, Daniel C and Issa) have responded to you that we are not keen to ditch Docker and compose.

daniellohrey-hai replied:
Nice, that reason makes sense, thanks for calling that out. Not sure if `There are many reasons to run Terraform locally as well as in CI.` is a typo for docker, but would like to clarify; is the plan to run terraform with docker in CI as well, or just locally?

What are you thoughts on the other proposed changes? Keen to get some automation going across Harrison.

schattingh replied:
I have responded here - https://github.com/harrison-ai/cookiecutter-terraform/pull/21#issuecomment-1350776798

comozo reviewed (changes requested):
Thanks for trying to improve our repo. I have raised few concerns. Happy to have a chat about it because I think those changes can have a big impact on a lot of folks.

daniellohrey-hai commented:
@schattingh 

> Hi @daniellohrey-hai
> 
> As a starting point, can you please motivate why you are requesting each one of the changes in this PR? What is the reason for each change?
> 
> Secondly, my gut reaction to the public `actions-library` repo is that its too easy for something that shouldn't be public to make its way in there unintended. Not saying we need to remove it - will get back to you on it.

Changes:
- Added repo name to cookiecutter variables:
  - In case the repo name is different to the project slug
  - Did remove the default tag for Repo since it was a duplicate of project, but that doesn't really fit with this change; don't mind which way we go, but should be consistent between both
- Refactored tf directory:
  - Feels more intuitive/easier to read/update
  - Easier to manage shared/environment specific variables
  - Symlinks are gross imo
  - No duplicated config
  - Easy to pass files to tf commands dynamically
  - Variables are best practice over locals for dynamic config
- Removed .env files:
  - Adds extra, unneeded config that isn't commited to source control
  - Difficult to work with in workflow
  - Had to manage/update when working locally
- Removed docker-compose to run terraform (also see discussion with Issa^):
  - Docker slows down the pipeline, and if it's not used in workflows it defeats its purpose (as outlined in 3 musketeers)
  - Some people have trouble running docker on the M1 chips
  - If anyone has trouble running terraform locally I'm happy to add in (the option to use) docker, but either way, I don't think we need compose
  - Adds extra steps/complexity/clutter to the repo without any benefit (to running terraform)
- Updated Makefile:
  - A lot of the targets are similar; updated to pass ENV=prod to make to avoid duplication
  - Not sure we want to use it in workflows?
  - Will re-add targets for things like fmt and validate (which no longer needs to be a script)
- Added GitHub Actions workflow (that references a reusable template):
  - The main reason behind the changes
  - Adds simple automation to our repos
  - Lays a foundation to discuss what we want our workflows to look like and how we can get there

If you want to read a little bit more about some of the above you can check out this page on confluence:
https://harrison-ai.atlassian.net/wiki/spaces/HSWE/pages/1934491649/Terraform+Workflow+Structure+Discussion
Discussion is in comments

To your second point: Any reusable workflows for GHA have to either live in a public repo or in the same repo as the calling workflow (which kind of defeats the purpose), so if we're using actions it essentially needs to be public.

Happy to discuss any points, not super attached to anything, just want to reach a point of agreement we're all happy with going forward.

rfk commented:
Thanks for the discussion here, all! I think there are a couple of threads in here that it would be great to discuss individually - for example, I also find the current `.env` approach quite fiddly and TBH a bit of a footgun, and would be keen to explore alternatives. 

Having it all in one big PR makes it hard to have those discussions, though. A good way to progress here may be to pick one thing (unsurprisingly, I propose ".env files" 😅) and kick around a few ideas in slack.

schattingh commented:
Thanks @daniellohrey-hai .  This PR is a no from me I'm afraid.  More than happy for other team members to chime in with their opinions.  It's a no for a number of reasons:

- The changes suggested are far reaching and radically change the way we work.  The blast radius and tech debt we'd accrue by making these changes are significant and would require significant hours and cost to eliminate.  This PR was described to me today by someone I respect as "nuclear".
- I haven't seen any evidence that these suggested changes would benefit us or make us work better, faster or smarter, despite specifically asking for this.  Describing aspects as "gross", are (in my opinion) just that; personal opinion, and not technical deficiencies.
- We have a number of questions as to how the suggested single directory would work in practice, and it would need a fair bit of testing before/if it were to be adopted.
- The linked public Github Action is incomplete and isn't suitable for use in most of our repos where multiple Terraform stacks exist.
- Claims like "Some people have trouble running docker on the M1 chips".  In the context of this discussion, this is unsubstantiated.

If you wish to tackle this again, (and I'm not discouraging you from doing so), here are a few unsolicited suggestions on how to land this better next time around:

- Speak to people first;  Find out if they are actually unhappy or dissatisfied with the current way of working.
- Find out the reasons why things are done a certain way before trying to change them.
- Bring us along the journey.
- Back up your suggested changes with technical benefits.  "I think we should use a single terraform directory for x, y, z reasons.  The existing directory structure of symlinks is inferior because a, b, c reasons, etc, etc."
- Demonstrate how your suggested changes can improve how we work, and ultimately the quality of what we produce.

In the spirit of transparency, I'm about to land two PR's in this repo in the next little while.  The first will remove the use of the `profile` argument to remove hard coded profile names, and the second will be to land the Github Actions workflow I shared with Issa a few weeks back.   I note @rfk  also suggested removing 3Musketeers and I'd welcome you to submit a PR to that effect.

Lastly, I'm also going to ask you to please make the `harrison-ai/actions-library` repo private, or indicate that the re-usable action it contains is not ready for production use.  In its current form, it indicates to the wider team that it has been formally reviewed and approved for re-use.  That is not the case.
